### PR TITLE
Inflate on demand & new listeners

### PR DESCRIPTION
**Pull request by pixsight**

### CODE REVIEW

File: sample/src/main/java/io/pixsight/sample/scenemanager/SampleActivity.kt [hunk 1]
```diff
@@ -21,50 +21,67 @@ import kotlinx.android.synthetic.main.spinner.*
...
+            R.id.sample_hide -> {
+                SceneManager.scene(this, SAMPLE_HIDDEN, true) // load a scene with a button so you can come back
+                SceneManager.hide(this, true) // all is now hidden
+                SceneManager.restore(this, true) // restore to SAMPLE_HIDDEN
+            }
+            else -> throw IllegalArgumentException("Nope")
         }
-        else -> throw IllegalArgumentException("Nope")
+
+        bindListeners()
```
Comment (tperraut) [inline sample/src/main/java/io/pixsight/sample/scenemanager/SampleActivity.kt:84]: Listeners need to be rebind on every button clicked?

File: sample/src/main/java/io/pixsight/sample/scenemanager/SampleBindScenesActivity.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_no_annotations_sample)
+        SceneManager.create(
+            SceneCreator.with(this)
+                .animation(SceneAnimations.TRANSLATE_X)
+        )
+
+        activity_no_annotations_sample_main_content.setOnClickListener(this)
```
Comment (tperraut) [inline sample/src/main/java/io/pixsight/sample/scenemanager/SampleBindScenesActivity.kt:42]: You did not use the same bindListeners pattern?

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneListener.kt [hunk 3]
```diff
@@ -7,18 +7,30 @@ package io.pixsight.scenemanager
...
-     * Note that this method is not called if there is no current scene set
+     * @param sceneId the new scene that is displayed to the user.
+     */
+    fun onSceneHiding(sceneId: Int) = Unit
+
+    /**
+     * Triggered when the out animation of the given sceneId finish
      *
-     * @param sceneId the scene that has been hidden.
+     * @param sceneId the new scene that is displayed to the user.
```
Comment (tperraut) [inline scenemanager/src/main/java/io/pixsight/scenemanager/SceneListener.kt]: This should return the scene that has finish its animation so the hidden scene (the same as onSceneHiding) no? If so, the old doc was correct for this case.

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt [hunk 4]
```diff
@@ -412,21 +451,53 @@ object SceneManager {
...
-        meta.sceneAnimationAdapter
-            .doChangeScene(scenesIdsToViews, meta.scenesParams, sceneId, animate)
-        notifyListener(meta.currentSceneId, sceneId, meta.listener)
-        meta.currentSceneId = sceneId
+        // inflate on demand
+        if (meta.inflateOnDemand) {
+            val currentSceneViews = scenesIdsToViews.get(sceneId)
+            val inflatedViews = ArrayList<Pair<Int, View>>()
+            currentSceneViews?.forEach {
+                if (it is InflateOnDemandLayout) {
```
Comment (tperraut) [inline scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt]: Could we name the it ?

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt [hunk 5]
```diff
@@ -412,21 +451,53 @@ object SceneManager {
...
+                    // The animation adapter may want to change the default attribute
+                    // to avoid the view from blinking. (before doChangeScene is called)
+                    meta.sceneAnimationAdapter.onViewInflatedOnDemand(sceneId, view)
+                    inflatedViews.add(Pair(it.id, view))
+                }
+            }
+
+            scenesIdsToViews.forEach { _, sceneViews ->
+                inflatedViews.forEach { pair ->
+                    if (sceneViews.removeAll { it.id == pair.first }) {
```
Comment (tperraut) [inline scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt]: Same here, what are pair and it ?

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt [hunk 6]
```diff
@@ -412,21 +451,53 @@ object SceneManager {
...
+        scenesIdsToViews: SparseArray<MutableList<View>>,
         sceneId: Int,
-        listener: SceneListener?
-    ) {
-        listener ?: return // nothing to do if no listener
-        if (currentSceneId != Int.MIN_VALUE) {
-            listener.onSceneHidden(currentSceneId)
+        animate: Boolean) {
+        // Cancel all pending animations
+        scenesIdsToViews.forEach { _, value ->
```
Comment (tperraut) [inline scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt]: Here we could name value -> views and it -> v or view

File: sample/src/main/java/io/pixsight/sample/scenemanager/SampleInflateOnDemandActivity.kt [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+class SampleInflateOnDemandActivity : AppCompatActivity(), View.OnClickListener {
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_inflate_on_demand_sample)
+        SceneManager.create(
+            SceneCreator.with(this)
+                .first(Scene.MAIN)
+                .animation(SceneAnimations.TRANSLATE_X)
+                .listener(object : SceneListener {
```
Comment (tperraut) [inline sample/src/main/java/io/pixsight/sample/scenemanager/SampleInflateOnDemandActivity.kt:45]: Did you check if all functions of the listener are triggered only once for one scene change?

File: scenemanager/src/main/java/io/pixsight/scenemanager/InflateOnDemandLayout.kt [hunk 8]
```diff
@@ -0,0 +1,82 @@
+package io.pixsight.scenemanager
```
Comment (tperraut) [inline scenemanager/src/main/java/io/pixsight/scenemanager/InflateOnDemandLayout.kt:1]: Maybe we should add an additional example in the README for this feature

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt [hunk 9]
```diff
@@ -232,6 +235,19 @@ object SceneManager {
      * @param creator a [SceneCreator]
      */
     fun create(creator: SceneCreator) {
+        val setup = getBuildAnnotation(creator.reference)
+        val scenes = setup?.value
+
+        scenes?.forEach { scene ->
+            scene.viewIds.forEach {
+                creator.add(scene.scene, it)
```
Comment (tperraut) [inline scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt]: Change it to sceneId or id ?

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt [hunk 10]
```diff
@@ -240,14 +256,16 @@ object SceneManager {
...
                     creator.scenes,
-                    creator.listener
+                    creator.listener,
+                    creator.inflateOnDemand
                 )
             )
         )
-        if (creator.firstSceneId != -1) {
+
+        if (setup != null || creator.firstSceneId != -1) {
```
Comment (tperraut) [inline scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt]: Change the -1 for a constant or just initialize it to Scene.NONE

**Review by tperraut — CHANGES_REQUESTED** [review]
The overall is awesome and will improve the SceneManager like never !
I just point with comment some details that need response or change :)

File: sample/src/main/java/io/pixsight/sample/scenemanager/SampleActivity.kt (hunk 1 above)
Comment (pixsight) [reply sample/src/main/java/io/pixsight/sample/scenemanager/SampleActivity.kt:84]: Les vues n'existent pas avant le inflate on demand. C'est pourquoi je rebind.

File: sample/src/main/java/io/pixsight/sample/scenemanager/SampleBindScenesActivity.kt (hunk 2 above)
Comment (pixsight) [reply sample/src/main/java/io/pixsight/sample/scenemanager/SampleBindScenesActivity.kt:42]: non car il n'y a pas le inflateOnDemand a true sur cette activity

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneListener.kt (hunk 3 above)
Comment (pixsight) [reply scenemanager/src/main/java/io/pixsight/scenemanager/SceneListener.kt]: Fix

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt (hunk 4 above)
Comment (pixsight) [reply scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt]: Fix

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt (hunk 6 above)
Comment (pixsight) [reply scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt]: Fix

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt (hunk 9 above)
Comment (pixsight) [reply scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt]: Fix

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt (hunk 10 above)
Comment (pixsight) [reply scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt]: Fix

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt (hunk 5 above)
Comment (pixsight) [reply scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt]: Je remplace les InflateOnDemandLayout dans les metas par les vues qui viennent d'etre inflate.
J'ai ajouté un commentaire et renommé les variables.

File: sample/src/main/java/io/pixsight/sample/scenemanager/SampleInflateOnDemandActivity.kt (hunk 7 above)
Comment (pixsight) [reply sample/src/main/java/io/pixsight/sample/scenemanager/SampleInflateOnDemandActivity.kt:45]: Oui les events hiding et hidden sont unique pour chaque scene. J'ai ajouté un filtre pour envoyer uniquement les events de la new et last scene.

File: scenemanager/src/main/java/io/pixsight/scenemanager/InflateOnDemandLayout.kt (hunk 8 above)
Comment (pixsight) [reply scenemanager/src/main/java/io/pixsight/scenemanager/InflateOnDemandLayout.kt:1]: Oui on va revoir le README.

File: scenemanager/src/main/java/io/pixsight/scenemanager/animations/TranslateXAnimationAdapter.kt [hunk 11]
```diff
@@ -43,100 +39,157 @@ class TranslateXAnimationAdapter(
...
-                // right to left
-                doRightToLeft(isNewScene, views)
+            // Animate each view if needed
+            val isLeftToRight = scenePosition < lastScenePosition
+            // If a view is in both scene, then we shouldn't animate
+            val animatedViews = views.filter {
+                !(scenesIdsToViews[scenesParams.lastSceneId].contains(it) &&
+                        scenesIdsToViews[sceneId].contains(it))
+            }
+            animatedViews.forEach { view ->
```
Comment (tperraut) [inline scenemanager/src/main/java/io/pixsight/scenemanager/animations/TranslateXAnimationAdapter.kt:67]: You can chain filter with onEach operator

Comment (pixsight) [reply scenemanager/src/main/java/io/pixsight/scenemanager/animations/TranslateXAnimationAdapter.kt:67]: je n'ai pas chain car j'ai besoin du dernier element pour **animatedViews.last()**

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt [hunk 12]
```diff
@@ -412,21 +451,90 @@ object SceneManager {
...
+        }
+    }
+
+    private fun createInternalListener(newSceneId: Int,
+                                       lastSceneId: Int,
+                                       listener: SceneListener?): SceneListener? {
+        listener ?: return null
+        return object : SceneListener {
+            override fun onSceneHiding(sceneId: Int) {
+                if (sceneId == newSceneId || sceneId == lastSceneId) {
```
Comment (tperraut) [inline scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt]: Shouldn't be only `sceneId == lastSceneId` ?

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt [hunk 13]
```diff
@@ -412,21 +451,90 @@ object SceneManager {
...
+                    listener.onSceneHiding(sceneId)
+                }
+            }
+
+            override fun onSceneHidden(sceneId: Int) {
+                if (sceneId == newSceneId || sceneId == lastSceneId) {
```
Comment (tperraut) [inline scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt]: Shouldn't be only `sceneId == lastSceneId` ?

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt [hunk 14]
```diff
@@ -412,21 +451,90 @@ object SceneManager {
...
+                    listener.onSceneHidden(sceneId)
+                }
+            }
+
+            override fun onSceneDisplaying(sceneId: Int) {
+                if (sceneId == newSceneId || sceneId == lastSceneId) {
```
Comment (tperraut) [inline scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt]: Shouldn't be only `sceneId == newSceneId ` ?

File: scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt [hunk 15]
```diff
@@ -412,21 +451,90 @@ object SceneManager {
...
+                    listener.onSceneDisplaying(sceneId)
+                }
+            }
+
+            override fun onSceneDisplayed(sceneId: Int) {
+                if (sceneId == newSceneId || sceneId == lastSceneId) {
```
Comment (tperraut) [inline scenemanager/src/main/java/io/pixsight/scenemanager/SceneManager.kt]: Shouldn't be only `sceneId == newSceneId` ?

**Review by tperraut — APPROVED** [review]

**Pull request merged**